### PR TITLE
Update Bevy to 0.12

### PR DESCRIPTION
**Pull request by sphw**
This PR updates the core plugin to Bevy 0.12. ~~Some examples are not compiling since they rely on the `bevy_spectator` plugin, which hasn't been updated yet.~~ 

UPDATE: I've now updated `bevy_spectator` to the latest version, and all the examples seem to build

### CODE REVIEW

**Comment by VirxEC** [comment]
This branch works for me on bevy 0.12 :+1:

File: src/pipeline.rs [hunk 1]
```diff
@@ -142,8 +142,8 @@ impl Plugin for AtmospherePipelinePlugin {
             .add_systems(
                 Render,
                 (
-                    Events::<AtmosphereUpdateEvent>::update_system.in_set(RenderSet::Prepare),
-                    prepare_atmosphere_assets.in_set(PrepareAssetSet::PostAssetPrepare),
+                    //Events::<AtmosphereUpdateEvent>::update_system.in_set(RenderSet::Prepare),
```
Comment (Shatur) [inline src/pipeline.rs:145]: Why this one is commented out?

File: src/skybox.rs [hunk 2]
```diff
@@ -15,11 +15,11 @@ use bevy::{
 pub struct AtmosphereSkyBoxMaterial(pub Handle<SkyBoxMaterial>);
 
 /// The `Handle` for the shader for the [`SkyBoxMaterial`].
-pub const ATMOSPHERE_SKYBOX_SHADER_HANDLE: HandleUntyped =
-    HandleUntyped::weak_from_u64(Shader::TYPE_UUID, 4511926918914205353);
+pub const ATMOSPHERE_SKYBOX_SHADER_HANDLE: Handle<Shader> =
```
Comment (Shatur) [inline src/skybox.rs:18]: Could you use newly added `embedded_asset!` instead?

File: macros/src/model.rs [hunk 3]
```diff
@@ -143,7 +143,7 @@ pub fn derive_atmospheric(ast: syn::DeriveInput) -> Result<TokenStream> {
         ShaderPathType::Internal(s) => quote! {
             {
                 use bevy::reflect::TypeUuid;
-                let handle = #asset_path::HandleUntyped::weak_from_u64(#render_path::render_resource::Shader::TYPE_UUID, #id);
+                let handle: #asset_path::Handle<Shader> = #asset_path::Handle::weak_from_u128(#id as u128);
```
Comment (Shatur) [inline macros/src/model.rs:146]: UUID is no longer necessary. Could you just use `embed_asset` instead and remove it from the macro?

File: src/pipeline.rs [hunk 4]
```diff
@@ -394,14 +394,11 @@ fn queue_atmosphere_bind_group(
...
-        layout: &image_bind_group_layout.0,
-        entries: &[BindGroupEntry {
-            binding: 0,
-            resource: BindingResource::TextureView(view),
-        }],
-    });
+    let image_bind_group = render_device.create_bind_group(
+        "bevy_atmosphere_image_bind_group",
+        &image_bind_group_layout.0,
+        &BindGroupEntries::sequential((BindingResource::TextureView(view),)),
```
Comment (Shatur) [inline src/pipeline.rs:400]:
```suggestion
        &BindGroupEntries::single(BindingResource::TextureView(view)),
```

**Review by Shatur — CHANGES_REQUESTED** [review]
Left a few suggestions and I will merge it and draft a new release.

File: src/skybox.rs (hunk 2 above)
Comment (sphw) [reply src/skybox.rs:18]: I just gave that a go, but am running into a panic from an unwrap with `embedded_asset!`, when used like this `        embedded_asset!(app, "shaders/skybox.wgsl");` in `plugin.rs`. I will take a deeper look tomorrow and hopefully push up a fix

Comment (sphw) [reply src/skybox.rs:18]: And just as a posted, I figured it out. Bevy does some slightly strange path manipulation using `file!()`, and it assumes that `file!()` will always return with a leading `/src/`, but that doesn't seem to be the case all of the time. Luckily you can work around this issue with: `embedded_asset!(app, "src/", "shaders/skybox.wgsl");`

Comment (Shatur) [reply src/skybox.rs:18]: Great!

**Comment by Shatur** [comment]
Thanks! But why `src/` is necessary? Isn't it `/src/` by default?

**Comment by sphw** [comment]
> Thanks! But why `src/` is necessary? Isn't it `/src/` by default?

The root issue is described in this ticket: https://github.com/bevyengine/bevy/issues/10377. Internally `embedded_asset` calls `file!()` which if it is in an nested workspace returns something like `crate_path/src/foo.rs`, but if it is in a normal crate (not in a workspace) it returns `src/foo.rs`. The default `/src/` assumes that the path will always contain that beginning `/`. My quick workaround was to just use `src/`.

**Review by Shatur — APPROVED** [review]

**Comment by Shatur** [comment]
Thank you a lot!

**Comment by Shatur** [comment]
@sphw this is odd, but when I trying `cargo publish` it fails with the following message:
```
error: couldn't read /mnt/Files/Code/bevy_atmosphere/target/package/bevy_atmosphere-0.8.0/src/../shaders/nishita.wgsl: No such file or directory (os error 2)
 --> src/collection/nishita.rs:7:10
  |
7 | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
  |          ^^^^^^^^^^^
  |
  = note: this error originates in the macro `include_str` which comes from the expansion of the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

error: couldn't read /mnt/Files/Code/bevy_atmosphere/target/package/bevy_atmosphere-0.8.0/src/../shaders/gradient.wgsl: No such file or directory (os error 2)
 --> src/collection/gradient.rs:7:10
  |
7 | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
  |          ^^^^^^^^^^^
  |
  = note: this error originates in the macro `include_str` which comes from the expansion of the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `HandleUntyped` in `asset`
 --> src/collection/nishita.rs:7:10
  |
7 | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
  |          ^^^^^^^^^^^ could not find `HandleUntyped` in `asset`
  |
  = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0433]: failed to resolve: could not find `HandleUntyped` in `asset`
 --> src/collection/gradient.rs:7:10
  |
7 | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
  |          ^^^^^^^^^^^ could not find `HandleUntyped` in `asset`
  |
  = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `std::option::Option<&str>: From<&BindGroupDescriptor<'_>>` is not satisfied
  --> src/collection/nishita.rs:7:10
   |
7  | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
   |          ^^^^^^^^^^^ the trait `From<&BindGroupDescriptor<'_>>` is not implemented for `std::option::Option<&str>`
   |
   = help: the following other types implement trait `From<T>`:
             <std::option::Option<Level> as From<LevelFilter>>
             <std::option::Option<bevy::utils::tracing::Id> as From<Span>>
             <std::option::Option<bevy::utils::tracing::Id> as From<tracing_core::span::Current>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a Span>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a EnteredSpan>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a bevy::utils::tracing::Id>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a tracing_core::span::Current>>
             <std::option::Option<&'a bevy::utils::tracing::Id> as From<&'a Span>>
           and 7 others
   = note: required for `&BindGroupDescriptor<'_>` to implement `Into<std::option::Option<&str>>`
note: required by a bound in `RenderDevice::create_bind_group`
  --> /home/gena/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_render-0.12.0/src/renderer/render_device.rs:89:21
   |
87 |     pub fn create_bind_group<'a>(
   |            ----------------- required by a bound in this associated function
88 |         &self,
89 |         label: impl Into<wgpu::Label<'a>>,
   |                     ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RenderDevice::create_bind_group`
   = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0061]: this method takes 3 arguments but 1 argument was supplied
  --> src/collection/nishita.rs:7:10
   |
7  | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
   |          ^^^^^^^^^^^ two arguments of type `&BindGroupLayout` and `&[BindGroupEntry<'_>]` are missing
   |
note: method defined here
  --> /home/gena/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_render-0.12.0/src/renderer/render_device.rs:87:12
   |
87 |     pub fn create_bind_group<'a>(
   |            ^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)
help: provide the arguments
   |
7  | #[derive(Atmospheric(Atmospheric, /* &BindGroupLayout */, /* &[BindGroupEntry<'_>] */), ShaderType, Reflect, Debug, Clone)]
   |                     ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0599]: no associated item named `TYPE_UUID` found for struct `bevy::prelude::Shader` in the current scope
 --> src/collection/nishita.rs:7:10
  |
7 | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
  |          ^^^^^^^^^^^ associated item not found in `Shader`
  |
  = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `std::option::Option<&str>: From<&BindGroupDescriptor<'_>>` is not satisfied
  --> src/collection/gradient.rs:7:10
   |
7  | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
   |          ^^^^^^^^^^^ the trait `From<&BindGroupDescriptor<'_>>` is not implemented for `std::option::Option<&str>`
   |
   = help: the following other types implement trait `From<T>`:
             <std::option::Option<Level> as From<LevelFilter>>
             <std::option::Option<bevy::utils::tracing::Id> as From<Span>>
             <std::option::Option<bevy::utils::tracing::Id> as From<tracing_core::span::Current>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a Span>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a EnteredSpan>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a bevy::utils::tracing::Id>>
             <std::option::Option<bevy::utils::tracing::Id> as From<&'a tracing_core::span::Current>>
             <std::option::Option<&'a bevy::utils::tracing::Id> as From<&'a Span>>
           and 7 others
   = note: required for `&BindGroupDescriptor<'_>` to implement `Into<std::option::Option<&str>>`
note: required by a bound in `RenderDevice::create_bind_group`
  --> /home/gena/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_render-0.12.0/src/renderer/render_device.rs:89:21
   |
87 |     pub fn create_bind_group<'a>(
   |            ----------------- required by a bound in this associated function
88 |         &self,
89 |         label: impl Into<wgpu::Label<'a>>,
   |                     ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RenderDevice::create_bind_group`
   = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0061]: this method takes 3 arguments but 1 argument was supplied
  --> src/collection/gradient.rs:7:10
   |
7  | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
   |          ^^^^^^^^^^^ two arguments of type `&BindGroupLayout` and `&[BindGroupEntry<'_>]` are missing
   |
note: method defined here
  --> /home/gena/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_render-0.12.0/src/renderer/render_device.rs:87:12
   |
87 |     pub fn create_bind_group<'a>(
   |            ^^^^^^^^^^^^^^^^^
   = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)
help: provide the arguments
   |
7  | #[derive(Atmospheric(Atmospheric, /* &BindGroupLayout */, /* &[BindGroupEntry<'_>] */), ShaderType, Reflect, Debug, Clone)]
   |                     ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0599]: no associated item named `TYPE_UUID` found for struct `bevy::prelude::Shader` in the current scope
 --> src/collection/gradient.rs:7:10
  |
7 | #[derive(Atmospheric, ShaderType, Reflect, Debug, Clone)]
  |          ^^^^^^^^^^^ associated item not found in `Shader`
  |
  = note: this error originates in the derive macro `Atmospheric` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0061, E0277, E0433, E0599.
For more information about an error, try `rustc --explain E0061`.
error: could not compile `bevy_atmosphere` (lib) due to 10 previous errors
warning: build failed, waiting for other jobs to finish...
error: failed to verify package tarball
```

Do you have any idea why it might happen?
`cargo check --examples` works just fine...

**Comment by sphw** [comment]
> @sphw this is odd, but when I trying cargo publish it fails with the following message:

I believe this problem is caused by using an outdated version of the derive macro crate. I think you first need to publish the derive macro crate, then bump the version in the primary crate. I've used this tool https://github.com/pksunkara/cargo-workspaces in the past to fix this issue, but that was years ago so I'm not sure it's still the best option.

**Comment by Shatur** [comment]
Oh, right, thank you!
Not the original author, just helping maintain the crate.

**Pull request merged**